sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java

```diff
@@ -857,4 +861,50 @@ private SqlNode getSqlNodeFor(SqlInsert insert, int idx)
     }
     return src;
   }
+
+  @Override
+  protected void validateHavingClause(SqlSelect select)
+  {
+    super.validateHavingClause(select);
+    SqlNode having = select.getHaving();
+    if (containsOver(having)) {
+      throw buildCalciteContextException("Window functions are not allowed in HAVING", having);
+    }
+  }
+
+  private boolean containsOver(SqlNode having)
+  {
+    if (having == null) {
+      return false;
+    }
+    final Predicate<SqlCall> callPredicate = call -> call.getOperator() instanceof SqlOverOperator;
+    return containsCall(having, callPredicate);
+  }
+
+  // copy of SqlUtil#containsCall
+  /** Returns whether an AST tree contains a call that matches a given
+   * predicate. */
+  private static boolean containsCall(SqlNode node,
+      Predicate<SqlCall> callPredicate)
+  {
+    try {
+      SqlVisitor<Void> visitor =
```

Comment:
is it not possible to make this a boolean visitor? `true` indicates presence, and `false` and `null` indicating non-presence?

Reply:
I'm not sure if that would be beneficial as this is just a copy of `SqlUtil#containsCall` until we upgrade to a version which has the fix - after that it should be removed

Reply:
Ah ok, that makes sense then - thanks for the clarification.